tgintegration/collector.py

```diff
@@ -0,0 +1,174 @@
+import asyncio
+import logging
+from contextlib import asynccontextmanager
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from typing import AsyncContextManager, List, Optional, TYPE_CHECKING, Union
+
+from pyrogram.errors import RpcMcgetFail
+from pyrogram.filters import Filter
+from pyrogram.handlers import MessageHandler
+from pyrogram.types import Message
+
+if TYPE_CHECKING:
+    from tgintegration.botcontroller import BotController
+from tgintegration._handler_utils import add_handler_transient
+from tgintegration.containers.response import InvalidResponseError, Response
+from tgintegration.update_recorder import MessageRecorder
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class TimeoutSettings:
+    max_wait: float = 10
+    """
+    The maximum duration in seconds to wait for a response from the peer.
+    """
+
+    wait_consecutive: Optional[float] = None
+    """
+    The minimum duration in seconds to wait for another consecutive message from the peer after receiving a message.
+    This can cause the total duration to exceed the `max_wait` time.
+    """
+
+    raise_on_timeout: bool = False
+    """
+    Whether to raise an exception when a timeout occurs or to fail with a log message.
+    """
+
+
+class NotSet(object):
```

Comment:
Might be best to do a sentinel value in the way dataclasses does:
https://github.com/python/cpython/blob/master/Lib/dataclasses.py#L169

I.e.
```
class _NotSetType:
    pass
NOT_SET = _NotSetType()

...

class Expectation:
    min_messages: Union[int, _NotSetType] = NOT_SET
```